SPC Bin Counters and mmap

davideberius opened the pull request:
This is an update to the SPC code to add two major features: bin counters and the mmap interface.  Bin counters are designed to allow for more context in the counters so that one counter can have several sub-counters attached to it as bins with breakpoints to denote when each bin gets incremented.  These are used extensively for collective algorithm counters which count the number of times each collective algorithm is called and under which circumstances small/large communicator/message.  The mmap interface creates a shared file using mmap so tools can attach to this and read SPC values directly using an XML file to point out where the data for each counter resides.  This update also includes some bug fixes and quality of life improvements to the code.

Commit:
This is an update to the SPC code to add two major features:
bin counters and the mmap interface.  Bin counters are designed
to allow for more context in the counters so that one counter
can have several sub-counters attached to it as bins with
breakpoints to denote when each bin gets incremented.
These are used extensively for collective algorithm counters
which count the number of times each collective algorithm is
called and under which circumstances small/large communicator/
message.  The mmap interface creates a shared file using mmap
so tools can attach to this and read SPC values directly using
an XML file to point out where the data for each counter
resides.  This update also includes some bug fixes and quality
of life improvements to the code.

Signed-off-by: David Eberius <deberius@vols.utk.edu>

bosilca commented on ompi/mca/pml/ob1/pml_ob1_sendreq.c:
```diff
@@ -645,9 +645,11 @@ int mca_pml_ob1_send_request_start_prepare( mca_pml_ob1_send_request_t* sendreq,
-                    OMPI_SPC_BYTES_SENT_USER, OMPI_SPC_BYTES_SENT_MPI);
+
     if( OPAL_LIKELY( rc >= OPAL_SUCCESS ) ) {
+        /*SPC_USER_OR_MPI(sendreq->req_send.req_base.req_ompi.req_status.MPI_TAG, (ompi_spc_value_t)size,*/
```
remove

bosilca commented on ompi/mca/pml/ob1/pml_ob1_sendreq.h:
```diff
@@ -385,7 +386,9 @@ mca_pml_ob1_send_request_start_btl( mca_pml_ob1_send_request_t* sendreq,
     }
 #endif /* OPAL_CUDA_GDR_SUPPORT */
 
+    //SPC_BIN_RECORD(OMPI_SPC_P2P_MESSAGE_SIZE, size);
```
remove

davideberius replied:
I removed these temporarily to see if they had a major impact on overhead.  I have added them back in.

bosilca commented on ompi/mca/pml/ob1/pml_ob1_sendreq.h:
```diff
@@ -385,7 +386,9 @@ mca_pml_ob1_send_request_start_btl( mca_pml_ob1_send_request_t* sendreq,
 
+    //SPC_BIN_RECORD(OMPI_SPC_P2P_MESSAGE_SIZE, size);
     if( OPAL_LIKELY(size <= eager_limit) ) {
+        //SPC_RECORD(OMPI_SPC_EAGER_MESSAGES, 1);
```
remove

bosilca commented on ompi/mca/pml/ob1/pml_ob1_sendreq.h:
```diff
@@ -405,6 +408,7 @@ mca_pml_ob1_send_request_start_btl( mca_pml_ob1_send_request_t* sendreq,
             break;
         }
     } else {
+        //SPC_RECORD(OMPI_SPC_NOT_EAGER_MESSAGES, 1);
```
remove

bosilca commented on ompi/runtime/help-mpi-runtime.txt:
```diff
@@ -119,3 +119,25 @@ Aborting...
 MPI_T performance variables.  Your job will continue, but SPCs will be
 disabled for MPI_T.
+#
+[spc: /dev/shm failed]
```
Does this really need to be /dev/shm ?

davideberius replied:
I suppose it doesn't have to be /dev/shm, but that's the default right now.  I think we fall back to the working directory if it's unavailable.

bosilca commented on ompi/runtime/ompi_mpi_params.c:
```diff
@@ -76,7 +76,13 @@ uint32_t ompi_add_procs_cutoff = OMPI_ADD_PROCS_CUTOFF_DEFAULT;
 bool ompi_mpi_dynamics_enabled = true;
 
 char *ompi_mpi_spc_attach_string = NULL;
```
Can we protect these so that they dont appear in the library if SPC is not enabled ?

bosilca commented on ompi/runtime/ompi_mpi_params.c:
```diff
@@ -3,7 +3,7 @@
  * Copyright (c) 2004-2005 The Trustees of Indiana University and Indiana
```
I think I would move all the changes in this file in the ompi_spc.c and create a registration function that is called from ompi_mpi_params.c.

bosilca commented on ompi/runtime/ompi_spc.c:
```diff
@@ -247,26 +394,339 @@ static int ompi_spc_get_count(const struct mca_base_pvar_t *pvar, void *value, v
+     * ###############################################################################
+     */
+
+    int bytes_needed = PAGE_SIZE;
```
Really ? You need a while loop to compute the number of pages needed for certain amount of data ?

How about ```bytes_needed = PAGE_SIZE * ((data_size  + PAGE_SIZE - 1) % PAGE_SIZE);``` ?

davideberius replied:
Fixed

bosilca commented on ompi/runtime/ompi_spc.c:
```diff
@@ -339,86 +842,232 @@ void ompi_spc_init(void)
+    ompi_spc_value_t *bins;
+
+    /* Denoted unlikely because counters will often be turned off. */
+    if( OPAL_UNLIKELY(IS_SPC_BIT_SET(ompi_spc_attached_event, event_id)) ) {
```
Can we get this check outside the function and into the upper-level macro ? If we can, we can save the function call.

davideberius replied:
Ok, I've added the check into the macros.

bosilca commented on opal/mca/base/mca_base_pvar.c:
```diff
@@ -204,6 +204,7 @@ int mca_base_pvar_register (const char *project, const char *framework, const ch
     case MCA_BASE_PVAR_CLASS_STATE:
         /* states MUST be integers */
         if (MCA_BASE_VAR_TYPE_INT != type) {
+            printf("It's not an integer!\n");
```
oops ?

bosilca commented on opal/mca/base/mca_base_pvar.c:
```diff
@@ -204,6 +204,7 @@ int mca_base_pvar_register (const char *project, const char *framework, const ch
     case MCA_BASE_PVAR_CLASS_STATE:
```
remove all changes to this file.

bosilca commented on opal/mca/btl/portals4/configure.m4:
```diff
@@ -11,7 +11,7 @@
 # Copyright (c) 2004-2005 The Regents of the University of California.
```
This should not be here ? Did you missed a merge or maybe you rebase according to OMPI master and the ICLDisco master is slightly behind ?

bosilca commented on orte/mca/state/state.h:
```diff
@@ -3,6 +3,9 @@
  * Copyright (c) 2011-2015 Los Alamos National Security, LLC. All rights
```
There should be no changes here.

bosilca reviewed (changes requested):
Good job! Overall the patch looks good. There are some mishaps that I marked, and few pending questions but we're getting to the end of it. I do however, deplore a certain lack of documentation, and it will not be easy for follow up developers to continue the work here without that.

davideberius commented on ompi/mca/mtl/portals4/mtl_portals4.c:
```diff
@@ -9,7 +9,7 @@
  *                         University of Stuttgart.  All rights reserved.
```
I don't think I changed this.  I think ICLDisco master is behind OMPI master.

davideberius replied:
Hmmm, I'm not sure.  I don't ever remember touching this file.  I guess I could revert this so it doesn't show up.

bosilca commented on orte/mca/state/orted/state_orted.c:
```diff
@@ -90,6 +93,107 @@ static orte_state_cbfunc_t proc_callbacks[] = {
+    for(i = 0; i < node->num_procs; i++) {
+        if(NULL != (pptr = (orte_proc_t*)opal_pointer_array_get_item(node->procs, i))) {
+            if(fd[i] == -1) {
+                rc = sprintf(sm_file, "%s" OPAL_PATH_SEP "spc_data.%s.%d.%d", shm_dir,
```
I am no sure I understand what this striking similar code is doing in the 2 cases.

bosilca commented on ompi/runtime/ompi_spc.c:
```diff
@@ -188,19 +247,95 @@ static int ompi_spc_notify(mca_base_pvar_t *pvar, mca_base_pvar_event_t event, v
+                }
+
+                int rank = ompi_comm_rank(ompi_spc_comm), rc;
+                char filename[64];
```
64 might quickly become too short.

bosilca commented on ompi/runtime/ompi_spc.c:
```diff
@@ -188,19 +247,95 @@ static int ompi_spc_notify(mca_base_pvar_t *pvar, mca_base_pvar_event_t event, v
+                char filename[64];
+
+                if(ompi_mpi_spc_xml_string == NULL) {
+                    rc = sprintf(filename, "%s" OPAL_PATH_SEP "spc_data.%s.%d.%d.xml", shm_dir,
```
As you set the length for the filename string please use ```snprintf``` instead of ```sprintf```.

bosilca commented on ompi/runtime/ompi_spc.c:
```diff
@@ -188,19 +247,95 @@ static int ompi_spc_notify(mca_base_pvar_t *pvar, mca_base_pvar_event_t event, v
+            index = pvar->pvar_index - mpi_t_offset;
+            if(index < 0) {
+                char *shm_dir;
+                if(0 == access("/dev/shm", W_OK)) {
```
Can we use opal_shmem_mmap_backing_file_base_dir instead of the hardcoded /dev/shm ?

Commit:
This update introduces changes to how watermark counters
are handled, adds new counters for internal queues such
as counters for data usage and timers for out-of-sequence
messages, and adds example code and documentation for
SPCs.  The watermark counter update reduces function call
overhead by better merging the updates to the sentinel
value and the watermark value.  There are now three SPC
example codes, one for accessing SPCs through MPI_T, one
for accessing SPCs through the mmap interface, and one
for parsing SPC snapshot datafiles and producing a
heatmap plot of the counter value differences over time.
This also adds documentation for SPCs in the form of a
markdown file in the same directory as the SPC driver
code.

Signed-off-by: David Eberius <deberius@vols.utk.edu>

bosilca commented on ompi/mca/pml/ob1/pml_ob1_recvreq.c:
```diff
@@ -1367,6 +1374,13 @@ void mca_pml_ob1_recv_req_start(mca_pml_ob1_recv_request_t *req)
+        }
+        SPC_RECORD(OMPI_SPC_UNEXPECTED_QUEUE_DATA, -total_data);
 #endif
             SPC_RECORD(OMPI_SPC_UNEXPECTED_IN_QUEUE, -1);
```
alignment ?

bosilca commented on orte/mca/state/orted/state_orted.c:
```diff
@@ -2,6 +2,9 @@
  * Copyright (c) 2011-2017 Los Alamos National Security, LLC.
```
this need to be moved somewhere else as there is no ORTE directory anymore.

Commit:
This update reflects the recent removal of orte in
favor of prrte.  This requires an update to the
SPC driver code and documentation and necessitates
moving the SPC snapshot feature to another code
region which will be in a separate pull request.

Signed-off-by: David Eberius <deberius@vols.utk.edu>
